### PR TITLE
[#IP-386] migrating to fp-ts 2.x

### PR DESCRIPTION
**Pull request by AleDore**
This PR upgrades to major release of:
- fp-ts
- io-ts

### CODE REVIEW

**Review by michaeldisaro — COMMENTED** [review]
Cannot see particular issues, but I'll check better on monday for safety.

File: openapi/index.yaml [hunk 1]
```diff
@@ -62,13 +62,13 @@ securityDefinitions:
...
 definitions:
   Timestamp:
-    $ref: "https://raw.githubusercontent.com/pagopa/io-functions-commons/v17.3.0/openapi/definitions.yaml#/Timestamp"
+    $ref: "https://raw.githubusercontent.com/pagopa/io-functions-commons/v21.2.0/openapi/definitions.yaml#/Timestamp"
   ProblemJson:
-    $ref: "https://raw.githubusercontent.com/pagopa/io-functions-commons/v17.3.0/openapi/definitions.yaml#/ProblemJson"
+    $ref: "https://raw.githubusercontent.com/pagopa/io-functions-commons/v21.2.0/openapi/definitions.yaml#/ProblemJson"
   OtpCode:
-    $ref: "https://raw.githubusercontent.com/pagopa/io-functions-cgn/v0.2.1/openapi/index.yaml#/definitions/OtpCode"
+    $ref: "https://raw.githubusercontent.com/pagopa/io-functions-cgn/IP_383_upgrade_to_fp_ts_2/openapi/index.yaml#/definitions/OtpCode"
```
Comment (michaeldisaro) [inline openapi/index.yaml:69]: We should remember to link the correct release here.

File: openapi/index.yaml [hunk 2]
```diff
@@ -62,13 +62,13 @@ securityDefinitions:
...
   Otp:
-    $ref: "https://raw.githubusercontent.com/pagopa/io-functions-cgn/v0.2.1/openapi/index.yaml#/definitions/Otp"
+    $ref: "https://raw.githubusercontent.com/pagopa/io-functions-cgn/IP_383_upgrade_to_fp_ts_2/openapi/index.yaml#/definitions/Otp"
```
Comment (michaeldisaro) [inline openapi/index.yaml:71]: ditto

File: utils/__tests__/redis_storage.test.ts [hunk 3]
```diff
@@ -15,70 +16,81 @@ const redisClientMock = {
...
-            value => expect(value).toEqual(aRedisValue)
-          )
-      )
-      .run();
+        O.fold(
+          () => fail(),
+          value => expect(value).toEqual(aRedisValue)
+        )
+      ),
+      TE.toUnion
```
Comment (michaeldisaro) [inline utils/__tests__/redis_storage.test.ts]: Te.toUnion is useless here, we are not assigning a result.

File: utils/__tests__/redis_storage.test.ts [hunk 4]
```diff
@@ -15,70 +16,81 @@ const redisClientMock = {
...
+    await pipe(
+      getTask(redisClientMock as any, aRedisKey),
+      TE.bimap(
         () => fail(),
-        maybeResult => expect(isNone(maybeResult)).toBeTruthy()
-      )
-      .run();
+        maybeResult => expect(O.isNone(maybeResult)).toBeTruthy()
+      ),
+      TE.toUnion
```
Comment (michaeldisaro) [inline utils/__tests__/redis_storage.test.ts]: Ditto.

File: utils/__tests__/redis_storage.test.ts [hunk 5]
```diff
@@ -15,70 +16,81 @@ const redisClientMock = {
...
-      .fold(
+    await pipe(
+      getTask(redisClientMock as any, aRedisKey),
+      TE.bimap(
         _ => expect(_).toBeDefined(),
         () => fail()
-      )
-      .run();
+      ),
+      TE.toUnion
```
Comment (michaeldisaro) [inline utils/__tests__/redis_storage.test.ts]: Ditto.

File: utils/__tests__/redis_storage.test.ts [hunk 6]
```diff
@@ -15,70 +16,81 @@ const redisClientMock = {
...
-      .fold(
+    await pipe(
+      deleteTask(redisClientMock as any, aRedisKey),
+      TE.bimap(
         () => fail(),
         del => expect(del).toBeTruthy()
-      )
-      .run();
+      ),
+      TE.toUnion
```
Comment (michaeldisaro) [inline utils/__tests__/redis_storage.test.ts]: Ditto.

File: utils/__tests__/redis_storage.test.ts [hunk 7]
```diff
@@ -15,70 +16,81 @@ const redisClientMock = {
...
-      .fold(
+    await pipe(
+      deleteTask(redisClientMock as any, aRedisKey),
+      TE.bimap(
         () => fail(),
         del => expect(del).toBeFalsy()
-      )
-      .run();
+      ),
+      TE.toUnion
```
Comment (michaeldisaro) [inline utils/__tests__/redis_storage.test.ts]: Ditto.

File: utils/__tests__/redis_storage.test.ts [hunk 8]
```diff
@@ -15,70 +16,81 @@ const redisClientMock = {
...
-      .fold(
+    await pipe(
+      deleteTask(redisClientMock as any, aRedisKey),
+      TE.bimap(
         _ => expect(_).toBeDefined(),
         () => fail()
-      )
-      .run();
+      ),
+      TE.toUnion
```
Comment (michaeldisaro) [inline utils/__tests__/redis_storage.test.ts]: Ditto.

File: utils/healthcheck.ts [hunk 9]
```diff
@@ -128,15 +133,14 @@ export const checkUrlHealth = (url: string): HealthCheck<"Url", true> =>
...
 export const checkApplicationHealth = (): HealthCheck<ProblemSource, true> =>
-  taskEither
-    .of<ReadonlyArray<HealthProblem<ProblemSource>>, void>(void 0)
-    .chain(_ => checkConfigHealth())
-    .chain(config =>
+  pipe(
+    TE.of(void 0),
+    TE.chain(_ => checkConfigHealth()),
+    TE.chainW(config =>
       // TODO: once we upgrade to fp-ts >= 1.19 we can use Validation to collect all errors, not just the first to happen
```
Comment (michaeldisaro) [inline utils/healthcheck.ts]: ~This comment is useless now.~
This comment should be updated.
```suggestion
      // TODO: Use Validation to collect all errors, not just the first to happen
```

**Review by michaeldisaro — CHANGES_REQUESTED** [review]
Apart minor changes LGTM.

File: utils/healthcheck.ts (hunk 9 above)
Comment (gquadrati) [reply utils/healthcheck.ts]: Actually no: even if we moved to v2, we didn't implement this feature yet in this project.

Comment (michaeldisaro) [reply utils/healthcheck.ts]: Right! I correct the assertion above.

Comment (gquadrati) [reply utils/healthcheck.ts]: As soon as we merge https://github.com/pagopa/io-functions-commons/pull/203 we can remove `healthcheck.ts` at all

Comment (michaeldisaro) [reply utils/healthcheck.ts]: And so... remove the comment. 😃 (...with the whole file)

File: ValidateOtp/handler.ts [hunk 10]
```diff
@@ -83,50 +78,68 @@ export type OtpResponseAndFiscalCode = t.TypeOf<
...
+): TE.TaskEither<Error, O.Option<OtpResponseAndFiscalCode>> =>
+  pipe(
+    getTask(redisClient, `${OTP_PREFIX}${otpCode}`),
+    TE.chain(
+      O.fold(
+        () => TE.of(O.none),
+        otpPayloadString =>
+          pipe(
+            otpPayloadString,
+            parse,
```
Comment (gquadrati) [inline ValidateOtp/handler.ts]: interesting

File: ValidateOtp/handler.ts [hunk 11]
```diff
@@ -83,50 +78,68 @@ export type OtpResponseAndFiscalCode = t.TypeOf<
...
+            E.mapLeft(E.toError),
+            TE.fromEither,
+            TE.chain(_ =>
+              pipe(
```
Comment (gquadrati) [inline ValidateOtp/handler.ts]: you can use `flow` instead

File: ValidateOtp/handler.ts [hunk 12]
```diff
@@ -83,50 +78,68 @@ export type OtpResponseAndFiscalCode = t.TypeOf<
...
-              e => new Error(`Cannot decode Otp Payload [${readableReport(e)}]`)
+): TE.TaskEither<Error, O.Option<OtpResponseAndFiscalCode>> =>
+  pipe(
+    getTask(redisClient, `${OTP_PREFIX}${otpCode}`),
+    TE.chain(
+      O.fold(
+        () => TE.of(O.none),
+        otpPayloadString =>
+          pipe(
+            otpPayloadString,
```
Comment (gquadrati) [inline ValidateOtp/handler.ts]: you can use `flow` instead

File: package.json [hunk 13]
```diff
@@ -52,24 +52,22 @@
...
+    "io-ts": "^2.2.16",
     "node-fetch": "^2.6.1",
     "redis": "^3.0.2",
     "redis-clustr": "^1.7.0",
     "winston": "^3.2.1"
   },
   "resolutions": {
-    "handlebars": "~4.5.3",
-    "fp-ts": "1.17.4"
+    "handlebars": "~4.5.3"
```
Comment (gquadrati) [inline package.json:71]: You could upgrade `@pagopa/openapi-codegen-ts` dev dependency too

File: openapi/index.yaml [hunk 14]
```diff
@@ -62,13 +62,13 @@ securityDefinitions:
 
 definitions:
   Timestamp:
-    $ref: "https://raw.githubusercontent.com/pagopa/io-functions-commons/v17.3.0/openapi/definitions.yaml#/Timestamp"
+    $ref: "https://raw.githubusercontent.com/pagopa/io-functions-commons/v21.2.0/openapi/definitions.yaml#/Timestamp"
```
Comment (gquadrati) [inline openapi/index.yaml:65]: To get rid of the need of upgrading fn-commons ref every time, you could move to `index.template.yaml` implementation, like what we did in [functions-app](https://github.com/pagopa/io-functions-app/pull/170)

File: utils/redis_storage.ts [hunk 15]
```diff
@@ -1,6 +1,7 @@
-import { Either, isLeft, left, right, toError } from "fp-ts/lib/Either";
-import { fromNullable, Option } from "fp-ts/lib/Option";
-import { fromEither, TaskEither, tryCatch } from "fp-ts/lib/TaskEither";
+import * as E from "fp-ts/lib/Either";
+import { pipe } from "fp-ts/lib/function";
+import * as O from "fp-ts/lib/Option";
+import * as TE from "fp-ts/lib/TaskEither";
 import { RedisClient } from "redis";
```
Comment (gquadrati) [inline utils/redis_storage.ts:5]: Most of the fuctions defined in `redis_storage.ts` are equivalent to the ones defined in `io-function-cgn`, and they seems o be quite general-purpose. 
Maybe we can think about moving them to a common library?

**Review by gquadrati — APPROVED** [review]
LGTM, apart minor changes

File: utils/redis_storage.ts (hunk 15 above)
Comment (AleDore) [reply utils/redis_storage.ts:5]: Absolutely yes, we have to refactor this when we move shared code into a common library 👍

File: openapi/index.yaml (hunk 14 above)
Comment (AleDore) [reply openapi/index.yaml:65]: Yep, I'll open another PR for this purpose :)

**Review by gquadrati — APPROVED** [review]
LGTM

**Review by michaeldisaro — APPROVED** [review]

**Pull request merged**